### PR TITLE
Release/eta to rsa builder

### PR DESCRIPTION
**Pull request by alexh-wediscover**
## Purpose

Script to pull out RSA ad copy in an account from the ETAs which already exist. Runs on MCCs or child accounts.

## Background

> For content release

## Callouts to review

> Script currently creates Google sheet regardless of whether there is anything to output, and just lets the user know it's safe to delete if empty. I think changing this would make the script a bit more convoluted, and not sure it's worth the effort

## Checklist

> [Required] You should review the points in this checklist before requesting a review.
> 
> Note: put **x** in each `[ ]` to mark as completed and use `~~` to cross out irrelevant points.

- [x] The PR message has been filled out appropriately.
- ~[ ] Any relevant CI processes are running successfully.~
- [x] Commit messages are well structured and explain what/why changes are made.
- [x] A well-structured readme exists that conforms to repo standards.
- [x] Docstrings have been used comprehensively and inline comments where needed.
- [x] Code has been structured with testability, readability, edge cases and errors in mind.
- ~[ ] Enviornmental differences have been implemented according to repo standards.~
- ~[ ] Appropriate testing has been added (unit > integration > end-to-end).~
- [x] Observability and monitoring has been considered.
- ~[ ] Information security has been considered.~
- ~[ ] _(If relevant)_ New infrastructure has been provisioned.~
- ~[ ] _(If relevant)_ New process alerting has been set up. ~

### CODE REVIEW

File: build_rsas_from_etas.gs [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+      AdsManagerApp.select(accountIterator.next());
+      
+      var accountName = AdsApp.currentAccount().getName();
+      var accountId = AdsApp.currentAccount().getCustomerId();
+      var accountNameAndId = accountName + " (" + accountId + ")";  
+      var successCount = 0;
+      
+      Logger.log("Looking in account " + accountNameAndId + " ...");
+      var assets = getAssets(accountName, accountId);
+      if (assets !== 'no_rsas') {
```
Comment (JaimeBadiola) [inline build_rsas_from_etas.gs:40]: Is this correct? `!==`

File: build_rsas_from_etas.gs [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+        var sheet = ss.insertSheet(accountNameAndId)
+        var maxNoOfAssets = getMaxAssets(assets);
+        writeToSheet(sheet, assets, maxNoOfAssets, accountName, accountId);
+        
+        successCount++;
```
Comment (JaimeBadiola) [inline build_rsas_from_etas.gs:45]: `++` is the same as python's `+=1` ?

File: build_rsas_from_etas.gs [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+    }
+    Logger.log("Script run complete. Google Sheet location: " + ss.getUrl());
+  }
+  
+  // If child account, process on that account only
+  else if (executionContext === 'client_account') {
+    var accountName = AdsApp.currentAccount().getName();
+    var accountId = AdsApp.currentAccount().getCustomerId();
+    var accountNameAndId = accountName + " (" + accountId + ")";
+    var sheet = ss.getSheets()[0].setName(accountNameAndId);
```
Comment (JaimeBadiola) [inline build_rsas_from_etas.gs:71]: Why don't you create the sheet only when you find RSAS? or delete the sheet automatically if no rsas are found? Could this potentially create hundreds of empty gsheets if nobody manually deletes them?

File: build_rsas_from_etas.gs [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+    Logger.log("Script run complete. Google Sheet location: " + ss.getUrl()); 
+
+  }
+}
+
+// ========= UTILITY FUNCTIONS ==========================================================================================
+
+// Determine the type of account in which the script is running
+function getExecutionContext() {
+    if (typeof AdsManagerApp != "undefined") {
```
Comment (JaimeBadiola) [inline build_rsas_from_etas.gs:93]: Is this the only case when an account is set as undefined?

File: build_rsas_from_etas.gs [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+  }
+  
+  // Set values and bolding
+  sheet.getRange(1, 1, 1, numColumns).setValues([headers]).setFontWeights([boldings]);
+  
+  return null;
+}
+
+// Set outputs for sheet
+function pushDataToSheet(sheet, groupedAssets, maxNoAssets, accountName, accountId) {
```
Comment (JaimeBadiola) [inline build_rsas_from_etas.gs:230]: This is nice! 
You might want to try next time to write a modular function that can be reused no matter what the structure of the table is. ie passing a list of dicts and just creates the table no matter the col names etc..

File: entity_rsa_checker/main.gs [hunk 6]
```diff
@@ -1,108 +1,309 @@
-/*
-    Name:        WeDiscover - RSA Checker, Google Ads Script
```
Comment (JaimeBadiola) [inline entity_rsa_checker/main.gs:2]: Is this doc a duplicate of the other document?

**Review by JaimeBadiola — CHANGES_REQUESTED** [review]

File: build_rsas_from_etas.gs (hunk 1 above)
Comment (alexh-wediscover) [reply build_rsas_from_etas.gs:40]: Yeah
In Javascript `==` means 'check value is the same regardless of type', so `10 == "10"` is `true`
There's also `===` means means 'check same value and same type' so `10 === "10"` is `false`

`!=` and `!==` are the negative equivalents. In general you want to use the `===` and `!==` as they're stricter.

File: build_rsas_from_etas.gs (hunk 2 above)
Comment (alexh-wediscover) [reply build_rsas_from_etas.gs:45]: Yep!

File: build_rsas_from_etas.gs (hunk 3 above)
Comment (alexh-wediscover) [reply build_rsas_from_etas.gs:71]: Yeah, this is a good point. I can't delete the sheet without including G Drive scope, but I will look at only creating the sheet if RSAs are found.

People will only run as a one off though, so even if empty sheets are made it shouldn't be too many

File: build_rsas_from_etas.gs (hunk 4 above)
Comment (alexh-wediscover) [reply build_rsas_from_etas.gs:93]: `typeof AdsManagerApp` returns `undefined` if the account the script is being run in a child account (i.e. not MCC), so this checks if the script is in an MCC or not.

Also interesting to note I've used `!= "undefined"` here, when it should really be `!== undefined`

File: build_rsas_from_etas.gs (hunk 5 above)
Comment (alexh-wediscover) [reply build_rsas_from_etas.gs:230]: Ah nice, that's a good point!

File: entity_rsa_checker/main.gs (hunk 6 above)
Comment (alexh-wediscover) [reply entity_rsa_checker/main.gs:2]: Hmm yes I did steal this, I'll update line 2

File: build_rsas_from_etas.gs (hunk 1 above)
Comment (alexh-wediscover) [reply build_rsas_from_etas.gs:40]: https://www.w3schools.com/js/js_comparisons.asp

File: build_rsas_from_etas.gs (hunk 3 above)
Comment (JaimeBadiola) [reply build_rsas_from_etas.gs:71]: Ok. Marking this as resolved and we can add this change in next version. @alexh-wediscover can you add to our asana?

File: build_rsas_from_etas.gs (hunk 5 above)
Comment (JaimeBadiola) [reply build_rsas_from_etas.gs:230]: Same as above, if you can add to asana for a future version

File: entity_rsa_checker/main.gs (hunk 6 above)
Comment (JaimeBadiola) [reply entity_rsa_checker/main.gs:2]: I meant all the script looks to be the same as the other file

Comment (alexh-wediscover) [reply entity_rsa_checker/main.gs:2]: Some is taken from [this script](https://github.com/we-discover/public/blob/master/entity_rsa_checker/main.gs), but that's just to get the ad group IDs where there are no RSAs. The rest of the functionality is new

Comment (alexh-wediscover) [reply entity_rsa_checker/main.gs]: Whoops, yes thank you for spotting that!! I'd accidentally overwritten it.

**Review by JaimeBadiola — APPROVED** [review]

**Pull request merged**